python/scripts/rfi_receiver/rfi_receiver.py

```diff
@@ -810,76 +836,63 @@ def rfi_zeroing():
     # Downtime of RFI zeroing
     downtime_m = app.config["solar_transit_downtime_m"]
     downtime_s = downtime_m * 60
-    half_downtime_s = 0.5 * downtime_s
-    minutes_in_day = 24 * 60
-
-    # Endpoint parameters
-    url = "http://csBfs:54323/rfi-zeroing-toggle"
-    headers = {"content-type": "application/json", "Accept-Charset": "UTF-8"}
+    half_window_s = 0.5 * downtime_s
 
     logger.info("RFI Solar Transit Toggle: Starting thread")
     while not InitialKotekanConnection:
         time.sleep(1)
     while True:
         # Wait until the correct UTC time of the solar transit at DRAO (deals with daylight savings time)
-        t_now = datetime.datetime.utcnow()
-        t_transit = ephemeris.solar_transit(t_now)[0]
-        t_diff = datetime.datetime.utcfromtimestamp(t_transit) - t_now
+        time_now = ephemeris.ensure_unix(datetime.datetime.utcnow())
 
-        time_to_transit_s = abs(t_diff.total_seconds())
+        # Get the *next* transit in the future
+        time_to_next_transit = ephemeris.solar_transit(time_now) - time_now
 
-        # Check if we are in the transit window, if so set downtime accordingly
-        if time_to_transit_s < half_downtime_s:
-            downtime_s = time_to_transit_s + half_downtime_s
-        elif time_to_transit_s > (minutes_in_day - (0.5 * downtime_m)) * 60:
-            downtime_s = minutes_in_day * 60 - time_to_transit_s
-        else:
+        # Get the *nearest* transit which we need to determine if we are still in the window
+        time_to_nearest_transit = (
+            ephemeris.solar_transit(time_now - 12 * 3600) - time_now
+        )
 
-            # Time until half_downtime_s before next solar transit in seconds
-            sleep_time_s = abs(time_to_transit_s - half_downtime_s)
+        # Check if we are within the current transit window and wait until the end of it
+        if abs(time_to_nearest_transit) < half_window_s:
+            set_rfi_zeroing(False)
+            downtime_s = half_window_s + time_to_nearest_transit
+        # Otherwise, we wait until the start of the next transit window
+        else:
+            set_rfi_zeroing(True)
+            downtime_s = time_to_next_transit - half_window_s
```

Comment:
Github is being annoying so I can't actually make a suggestion block big enough, but I agree with Andre that this end of thing method is still a bit funky. You should only need to set the state once. Something like this probably works:


```suggestion
       log.debug("Next solar transit ...")
       log.debug("Nearest solar transit ...")

        # Check if we are within the current transit window and wait until the end of it
        if abs(time_to_nearest_transit) < half_window_s:
            new_zeroing_state = False
            downtime_s = half_window_s + time_to_nearest_transit
            log.info("Within transit window, disabling zeroing and waiting until the end") 
        # Otherwise, we wait until the start of the next transit window
        else:
            new_zeroing_state = True
            downtime_s = time_to_next_transit - half_window_s
            log.info("Outside transit window, enabling zeroing and waiting until next window") 

        success = set_rfi_zeroing(new_zeroing_state)
        
        if not success:
            # Figure out what to do in here, probably sleep for a few seconds,
            # and then just continue, which will basically force it to try again
            pass
            
        time.sleep(downtime_s)
```
And then remove everything else below in the function